test/unit/gem_download_test.rb

```diff
@@ -0,0 +1,149 @@
+require 'test_helper'
+
+class GemDownloadTest < ActiveSupport::TestCase
+  setup do
+    create(:gem_download, count: 0)
+  end
+
+  context "update_count_by" do
+    should "not update if download doesnt exist" do
+      assert_nil GemDownload.update_count_by(1, rubygem_id: 1)
+    end
+
+    should "not update if download count is nil" do
+      GemDownload.create!(rubygem_id: 1, version_id: 0)
+      download = GemDownload.update_count_by(1, rubygem_id: 1)
+      assert_nil download.count
+    end
+
+    should "update the count" do
+      GemDownload.create!(rubygem_id: 1, version_id: 1, count: 0)
+      GemDownload.update_count_by(1, rubygem_id: 1, version_id: 1)
+
+      assert_equal 1, GemDownload.where(rubygem_id: 1, version_id: 1).first.count
+    end
+  end
+
+  uses_transaction :test_update_the_count_atomically
```

Comment:
This test cannot run in a transactional fixture block, because I use threads and need to use different connections to the DB. in order to simulate a bit of concurrency .